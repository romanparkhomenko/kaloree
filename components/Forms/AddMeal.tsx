import React, { useState } from 'react';

import { Card, CardBody, CardTitle, Container, Row, Col, Form, FormGroup, Input } from 'reactstrap';

const AddMeal: React.FC = () => {
  const [food, setFood] = useState('');
  const [foodCategory, setFoodCategory] = useState('');
  const [grams, setGrams] = useState('');
  const [ounces, setOunces] = useState('');
  const [calories, setCalories] = useState('');
  const [successMessage, setSuccessMessage] = useState('');
  const [errorMessage, setErrorMessage] = useState('');

  const submitData = async (e: React.SyntheticEvent) => {
    e.preventDefault();
    setSuccessMessage('');
    setErrorMessage('');
    try {
      const body = { food, foodCategory, grams, ounces, calories };
      await fetch(`/api/meal`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(body),
      })
        .then(res => res.json())
        .then(response => {
          console.info(response);
          setSuccessMessage(`Successfully logged ${food}`);
          setFood('');
          setFoodCategory('');
          setGrams('');
          setOunces('');
          setCalories('');

          setTimeout(() => {
            setSuccessMessage('');
          }, 5000);
        })
        .catch(error => {
          setErrorMessage('Whoops, something went wrong');
          console.info(error);
        });
    } catch (error) {
      setErrorMessage('Whoops, something went wrong');
      console.error(error);
    }
  };

  const inputGrams = grams => {
    const g = parseFloat(grams);
    const oz = (g / 28.35).toFixed(2);
    console.info(grams);
    setGrams(g.toString());
    setOunces(oz);
  };

  const inputOunces = ounces => {
    const oz = parseFloat(ounces);
    const g = (oz * 28.35).toFixed(2);
    setOunces(oz.toString());
    setGrams(g);
  };

  return (
    <>
      <Card className="card-stats mb-4 mb-xl-0">
        <CardBody>
          <Row>
            <div className="col">
              <Form>
                <Col lg="12" className="align-items-center justify-content-between row">
                  <h6 className="heading-small text-muted mb-4">Log a Meal</h6>
                  <span className="icon icon-shape bg-gradient-info text-white rounded-circle shadow">
                    <i className="fas fa-chart-bar" />
                  </span>
                </Col>
                <Row>
                  <Col lg="12">
                    <label className="form-control-label" htmlFor="food">
                      Food
                    </label>
                    <Input
                      className="form-control-alternative"
                      id="food"
                      placeholder="What did you eat?"
                      onChange={e => setFood(e.target.value)}
                      value={food}
                      type="text"
                    />
                  </Col>
                  <Col lg="12">
                    <label className="form-control-label" htmlFor="foodCategory">
                      Food Category
                    </label>
                    <Input
                      className="form-control-alternative"
                      id="foodCategory"
                      placeholder="Carbs, Fruit, Fats..."
                      onChange={e => setFoodCategory(e.target.value)}
                      value={foodCategory}
                      type="text"
                    />
                  </Col>
                  <Col lg="12">
                    <label className="form-control-label" htmlFor="grams">
                      Grams
                    </label>
                    <Input
                      className="form-control-alternative"
                      id="grams"
                      placeholder="Grams"
                      onChange={e => inputGrams(e.target.value)}
                      value={grams}
                      type="number"
                    />
                  </Col>
                  <Col lg="6">
                    <label className="form-control-label" htmlFor="ounces">
                      Ounces
                    </label>
                    <Input
                      className="form-control-alternative"
                      id="ounces"
                      placeholder="Ounces"
                      onChange={e => inputOunces(e.target.value)}
                      value={ounces}
                      type="number"
                    />
                  </Col>
                  <Col lg="6">
                    <label className="form-control-label" htmlFor="calories">
                      Calories
                    </label>
                    <Input
                      className="form-control-alternative"
                      id="calories"
                      placeholder="Calories"
                      onChange={e => setCalories(e.target.value)}
                      value={calories}
                      type="number"
                    />
                  </Col>
                </Row>
                <Row className="align-items-center justify-content-end">
                  <Col lg="12" className="mt-3">
                    <button
                      className="btn btn-info bg-gradient-info w-100"
                      onClick={e => submitData(e)}
                    >
                      Log My Meal
                    </button>
                    {successMessage && (
                      <p className="text-center font-weight-700 text-info">{successMessage}</p>
                    )}
                    {errorMessage && (
                      <p className="text-center font-weight-700 text-warning">{errorMessage}</p>
                    )}
                  </Col>
                </Row>
              </Form>
            </div>
          </Row>
        </CardBody>
      </Card>
    </>
  );
};

export default AddMeal;
