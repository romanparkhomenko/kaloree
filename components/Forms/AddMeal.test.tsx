import * as React from 'react';
import { mount, shallow } from 'enzyme';
import { render, fireEvent, waitFor, screen, act } from '@testing-library/react';
import renderer from 'react-test-renderer';
import AddMeal from './AddMeal';

describe('Add Meal Form', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<AddMeal />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('All inputs work and user can submit data', async () => {
    render(<AddMeal />);

    const foodInput = screen.getByPlaceholderText('What did you eat?');
    const categoryInput = screen.getByPlaceholderText('Carbs, Fruit, Fats...');
    const gramsInput = screen.getByPlaceholderText('Grams');
    const ouncesInput = screen.getByPlaceholderText('Ounces');
    const caloriesInput = screen.getByPlaceholderText('Calories');
    const submitButton = screen.getByText('Log My Meal');

    fireEvent.change(foodInput, { target: { value: 'Bananas' } });
    expect(foodInput).toHaveValue('Bananas');

    fireEvent.change(categoryInput, { target: { value: 'Fruit' } });
    expect(categoryInput).toHaveValue('Fruit');

    // Ensure conversion occurs between grams and ounces
    fireEvent.change(gramsInput, { target: { value: '28.35' } });
    expect(gramsInput).toHaveValue(28.35);
    expect(ouncesInput).toHaveValue(1.0);

    fireEvent.change(ouncesInput, { target: { value: '3.00' } });
    expect(ouncesInput).toHaveValue(3.0);
    expect(gramsInput).toHaveValue(85.05);

    fireEvent.change(caloriesInput, { target: { value: '300' } });
    expect(caloriesInput).toHaveValue(300);

    const fakeWeightResponse = { okay: 'weight logged' };

    global.fetch = jest.fn().mockImplementation(() => {
      return Promise.resolve({
        json: () => Promise.resolve(fakeWeightResponse),
      });
    });

    await act(async () => {
      fireEvent.click(submitButton);
    });

    // Confirm success message exists and values are reset
    expect(screen.getByText('Successfully logged Bananas')).toBeVisible();
    expect(foodInput).toHaveValue('');
    expect(categoryInput).toHaveValue('');
    expect(gramsInput).toHaveValue(null);
    expect(ouncesInput).toHaveValue(null);
    expect(caloriesInput).toHaveValue(null);
  });
});
