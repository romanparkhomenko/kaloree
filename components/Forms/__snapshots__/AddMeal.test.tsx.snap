// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add Meal Form renders correctly 1`] = `
<div
  className="card-stats mb-4 mb-xl-0 card"
>
  <div
    className="card-body"
  >
    <div
      className="row"
    >
      <div
        className="col"
      >
        <form
          className=""
        >
          <div
            className="align-items-center justify-content-between row col-lg-12"
          >
            <h6
              className="heading-small text-muted mb-4"
            >
              Log a Meal
            </h6>
            <span
              className="icon icon-shape bg-gradient-info text-white rounded-circle shadow"
            >
              <i
                className="fas fa-chart-bar"
              />
            </span>
          </div>
          <div
            className="row"
          >
            <div
              className="col-lg-12"
            >
              <label
                className="form-control-label"
                htmlFor="food"
              >
                Food
              </label>
              <input
                className="form-control-alternative form-control"
                id="food"
                onChange={[Function]}
                placeholder="What did you eat?"
                type="text"
                value=""
              />
            </div>
            <div
              className="col-lg-12"
            >
              <label
                className="form-control-label"
                htmlFor="foodCategory"
              >
                Food Category
              </label>
              <input
                className="form-control-alternative form-control"
                id="foodCategory"
                onChange={[Function]}
                placeholder="Carbs, Fruit, Fats..."
                type="text"
                value=""
              />
            </div>
            <div
              className="col-lg-12"
            >
              <label
                className="form-control-label"
                htmlFor="grams"
              >
                Grams
              </label>
              <input
                className="form-control-alternative form-control"
                id="grams"
                onChange={[Function]}
                placeholder="Grams"
                type="number"
                value=""
              />
            </div>
            <div
              className="col-lg-6"
            >
              <label
                className="form-control-label"
                htmlFor="ounces"
              >
                Ounces
              </label>
              <input
                className="form-control-alternative form-control"
                id="ounces"
                onChange={[Function]}
                placeholder="Ounces"
                type="number"
                value=""
              />
            </div>
            <div
              className="col-lg-6"
            >
              <label
                className="form-control-label"
                htmlFor="calories"
              >
                Calories
              </label>
              <input
                className="form-control-alternative form-control"
                id="calories"
                onChange={[Function]}
                placeholder="Calories"
                type="number"
                value=""
              />
            </div>
          </div>
          <div
            className="align-items-center justify-content-end row"
          >
            <div
              className="mt-3 col-lg-12"
            >
              <button
                className="btn btn-info bg-gradient-info w-100"
                onClick={[Function]}
              >
                Log My Meal
              </button>
              
              
            </div>
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;

exports[`AddMeal form renders 1`] = `
<div
  className="card-stats mb-4 mb-xl-0 card"
>
  <div
    className="card-body"
  >
    <div
      className="row"
    >
      <div
        className="col"
      >
        <form
          className=""
        >
          <div
            className="align-items-center justify-content-between row col-lg-12"
          >
            <h6
              className="heading-small text-muted mb-4"
            >
              Log a Meal
            </h6>
            <span
              className="icon icon-shape bg-gradient-info text-white rounded-circle shadow"
            >
              <i
                className="fas fa-chart-bar"
              />
            </span>
          </div>
          <div
            className="row"
          >
            <div
              className="col-lg-12"
            >
              <label
                className="form-control-label"
                htmlFor="food"
              >
                Food
              </label>
              <input
                className="form-control-alternative form-control"
                id="food"
                onChange={[Function]}
                placeholder="What did you eat?"
                type="text"
                value=""
              />
            </div>
            <div
              className="col-lg-12"
            >
              <label
                className="form-control-label"
                htmlFor="foodCategory"
              >
                Food Category
              </label>
              <input
                className="form-control-alternative form-control"
                id="foodCategory"
                onChange={[Function]}
                placeholder="Carbs, Fruit, Fats..."
                type="text"
                value=""
              />
            </div>
            <div
              className="col-lg-12"
            >
              <label
                className="form-control-label"
                htmlFor="grams"
              >
                Grams
              </label>
              <input
                className="form-control-alternative form-control"
                id="grams"
                onChange={[Function]}
                placeholder="Grams"
                type="number"
                value=""
              />
            </div>
            <div
              className="col-lg-6"
            >
              <label
                className="form-control-label"
                htmlFor="ounces"
              >
                Ounces
              </label>
              <input
                className="form-control-alternative form-control"
                id="ounces"
                onChange={[Function]}
                placeholder="Ounces"
                type="number"
                value=""
              />
            </div>
            <div
              className="col-lg-6"
            >
              <label
                className="form-control-label"
                htmlFor="calories"
              >
                Calories
              </label>
              <input
                className="form-control-alternative form-control"
                id="calories"
                onChange={[Function]}
                placeholder="Calories"
                type="number"
                value=""
              />
            </div>
          </div>
          <div
            className="align-items-center justify-content-end row"
          >
            <div
              className="mt-3 col-lg-12"
            >
              <button
                className="btn btn-info bg-gradient-info w-100"
                onClick={[Function]}
              >
                Log My Meal
              </button>
              
              
            </div>
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
